File: solutions/c/secret-handshake/1/secret_handshake.c
#include <stdlib.h>
#include <stdbool.h>
#include "secret_handshake.h"

/*
00001 = wink
00010 = double blink
00100 = close your eyes
01000 = jump

10000 = Reverse the order of the operations in the secret handshake.
*/

bool isBitSet(int flags, int bit);

const char **commands(size_t number) {
    const char ** result = malloc(10 * sizeof(char*));
    int counter = 0;
    
    result[0] = NULL;

    if (isBitSet(number, 1)) {
        result[counter] = "wink";
        counter++;
    }
    
    if (isBitSet(number, 2)) {
        result[counter] = "double blink";
        counter++;
    }

    if (isBitSet(number, 4)) {
        result[counter] = "close your eyes";
        counter++;
    }

    if (isBitSet(number, 8)) {
        result[counter] = "jump";
        counter++;
    }
    
    // reverse it
    if (isBitSet(number, 16)) {
        if (counter > 1) {
            for(int i = 0; i < counter/2; i++) {
                const char* temp = result[i];
                result[i] = result[counter -1 -i];
                result[counter -1 -i] = temp;
            }
        }
    }

    return result;
}

bool isBitSet(int flags, int bit) {
    return (flags & bit) == bit;
}

File: solutions/csharp/secret-handshake/1/SecretHandshake.cs
using System;
using System.Collections.Generic;

public static class SecretHandshake
{
    private static Dictionary<CommandTypes, string> _commands = new Dictionary<CommandTypes, string>() {
        {CommandTypes.Wink, "wink"},
        {CommandTypes.DoubleBlink, "double blink"},
        {CommandTypes.CloseYourEyes, "close your eyes"},
        {CommandTypes.Jump, "jump"},
        {CommandTypes.Reverse, "reverse"}
    };

    [Flags]
    private enum CommandTypes 
    {
        Wink = 1,
        DoubleBlink = 2,
        CloseYourEyes = 4,
        Jump = 8,
        Reverse = 16
    }

    public static string[] Commands(int commandValue)
    {
        var ret = new List<string>();
        var reverse = false;
        var mask = (CommandTypes)commandValue;

        foreach(var item in Enum.GetNames(typeof(CommandTypes)))
        {
            var flag = (CommandTypes)Enum.Parse(typeof(CommandTypes), item);
            if(mask.HasFlag(flag)) 
            {
                if (flag == CommandTypes.Reverse) reverse = true;
                else ret.Add(_commands[flag]);
            }
        }

        if (reverse) ret.Reverse();
        return ret.ToArray();
    }
}
